## Exercise 1 - Question 1
Let $x_{i,j}$ be $x$ amount of advertisement from company $i$ shown on section * $j$. 
* $x$ is numerical. 
* $i\ \epsilon\ \{1,2,3,4\}$, representing GEICO, Delta, T-Mobile, and Capital One.
* $j\ \epsilon\ \{1, 2\}$, representing Sports and National.

Assume the purpose of the analysis is to maximize The Washington Post's revenue.

Optimization function:
* maximize: $2.4\times(x_{1,1}\times2.5\%+x_{1,2}\times0.8\%+x_{2,1}\times2\%+x_{2,2}\times1\%+x_{3,1}\times1\%+x_{3,2}\times3\%+x_{4,1}\times1.5\%+x_{4,2}\times2\%)$

Constraints One: the number of page view needs to meet minimum criteria.
* $x_{1,1}\geq2000000$
* $x_{1,2}\geq1000000$
* $x_{2,1}\geq0$
* $x_{2,2}\geq1000000$
* $x_{2,1} + x_{2,2}\geq2000000$
* $x_{3,1}\geq1000000$
* $x_{3,2}\geq1000000$
* $x_{3,1} + x_{3,2}\geq3000000$
* $x_{4,1}\geq0$
* $x_{4,2}\geq0$
* $x_{4,1} + x_{4,2}\geq2000000$

Constraints Two: Total amount of page view can't exceed The Washington Post's number of visits.
* $x_{1,1}+x_{2,1}+x_{3,1}+x_{4,1}\leq6000000$
* $x_{1,2}+x_{2,2}+x_{3,2}+x_{4,2}\leq5000000$

## Exercise 1 - Question 2
(For analysis please refer to the next block.)

The maximum revenue is $523200 per day.

On Sports section, ads from GEICO, Delta, T-Mobile, and Capital One will be shown 3, 1, 1, and 1 million times, respectively.

On National section, ads from GEICO, Delta, T-Mobile, and Capital One will be shown 1, 1, 2, and 1 million times, respectively.

In [75]:
import math
import numpy
import cvxpy as cvx


# Definition of the Variables
x11 = cvx.Variable(1, integer=True)# Geico Sports
x12 = cvx.Variable(1, integer=True)# Geico National
x21 = cvx.Variable(1, integer=True)# Delta Sports
x22 = cvx.Variable(1, integer=True)# Delta National
x31 = cvx.Variable(1, integer=True)# T-Mobile Sports
x32 = cvx.Variable(1, integer=True)# T-Mobile National
x41 = cvx.Variable(1, integer=True)# Capital One Sports
x42 = cvx.Variable(1, integer=True)# Capital One National

# Constraints
c1 = x11>=2000000 
c2 = x12>=1000000 
c3 = x21>=0
c4 = x22>=1000000
c5 = x21+x22>=2000000
c6 = x31>=1000000
c7 = x32>=1000000
c8 = x31+x32>=3000000
c9 = x41>=0
c10 = x42>=0
c11 = x41+x42>=2000000
c12 = x11+x21+x31+x41<=6000000
c13 = x12+x22+x32+x42<=5000000
con=[c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,c11,c12,c13]

#objective, k is expected revenue per ad
k11 = 2.5/100
k12 = 0.8/100
k21 = 2.0/100
k22 = 1.0/100
k31 = 1.0/100
k32 = 3.0/100
k41 = 1.5/100
k42 = 2.0/100
OF = (k11*x11 + k12*x12+ k21*x21 + k22*x22 + k31*x31+ k32*x32 + k41*x41 + k42*x42)*2.4
objective = cvx.Maximize(OF)

#solving
prob = cvx.Problem(objective, con)
result = prob.solve()

print('optimal revenue')
print(prob.value)

print('Allocation to Sports')
print(x11.value)
print(x21.value)
print(x31.value)
print(x41.value)

print('Allocation to National')
print(x12.value)
print(x22.value)
print(x32.value)
print(x42.value)

optimal revenue
523200.0
Allocation to Sports
[3000000.]
[1000000.]
[1000000.]
[1000000.]
Allocation to National
[1000000.]
[1000000.]
[2000000.]
[1000000.]


## Exercise 1 - Question 3

In [76]:
# Define alternative k, the expected click through rate for the sensitivity analysis
alt_CTR = {"k11_1":2, "k11_2":3, "k12_1":0.1, "k12_2":1.1, "k21_1":1.5, "k21_2":2.5, "k22_1":0.5, "k22_2":1.5, "k31_1":0.5, "k31_2":1.5, "k32_1":2.5, "k32_2":3.5, "k41_1":1.5, "k41_2":2.0, "k42_1":1.5, "k42_2":2.5}
for i in alt_CTR:
  alt_CTR[i] = alt_CTR[i]/100.0

# Apply each alternative CTR and estimate the new optimal revenue
for i in range(0, 16):
  k = [k11, k12, k21, k22, k31, k32, k41, k42]
  k[int(i/2)] = list(alt_CTR.values())[i]
  alt_OF = (k[0]*x11 + k[1]*x12+ k[2]*x21 + k[3]*x22 + k[4]*x31+ k[5]*x32 + k[6]*x41 + k[7]*x42)*2.4
  objective = cvx.Maximize(alt_OF)
  prob = cvx.Problem(objective, con)
  result = prob.solve()
  print(f'optimal revenue when {list(alt_CTR.items())[i]}: ', prob.value)

optimal revenue when ('k11_1', 0.02):  499200.0
optimal revenue when ('k11_2', 0.03):  559200.0
optimal revenue when ('k12_1', 0.001):  506400.0
optimal revenue when ('k12_2', 0.011000000000000001):  530400.0
optimal revenue when ('k21_1', 0.015):  511200.0
optimal revenue when ('k21_2', 0.025):  535200.0
optimal revenue when ('k22_1', 0.005):  511200.0
optimal revenue when ('k22_2', 0.015):  535200.0
optimal revenue when ('k31_1', 0.005):  511200.0
optimal revenue when ('k31_2', 0.015):  535200.0
optimal revenue when ('k32_1', 0.025):  499200.0
optimal revenue when ('k32_2', 0.035):  559200.0
optimal revenue when ('k41_1', 0.015):  523200.0
optimal revenue when ('k41_2', 0.02):  547200.0
optimal revenue when ('k42_1', 0.015):  523200.0
optimal revenue when ('k42_2', 0.025):  535200.0


## Exercise 1 - Question 4
Variations of CTR for some advertisers are sensitive. For example, when GEICO's CTR in Sports grows from 2.5% to 3%, the optimal revenue will increase from \$523200 to $\$$559200. T-Mobile's CTR on the National section is sensitive too. If CTR drops from 3% to 2.5%, the optimal revenue will decrease from \$523200 to $\$$499200. Other ads, by comparison, are less sensitive with significantly smaller variations in revenue when CTR changes. When pursuing maximum revenue, we should pay more attention to GEOCO and T-Mobile, make sure their ads are attractive.

## Exercise 2 - Question 1
Decision variables: $x_{i,j}$.

Let $x_{i,j}$ be plant $i$ makes $x$ amount of $j$ products. 
* $x$ is numerical. 
* $i\  \epsilon\ \{1,2\}$, representing Plant A and Plant B.
* $j\  \epsilon\ \{1, 2\}$, representing Standard and Deluxe.

Goal: Assume to maximize the profit.

>Objective function for Plant A:
* maximize: $10\times x_{1,1}+15\times x_{1,2}$

>Objective function for Plant B:
* maximize: $10\times x_{2,1}+15\times x_{2,2}$

Constraints One: the time to produce products can't exceed the plant's capacity.

>For Plant A:
* $x_{1,1}\times4+x_{1,2}\times2\leq80$
* $x_{1,1}\times2+x_{1,2}\times5\leq60$
* $x_{1,1} \geq0$
* $x_{1,2} \geq0$

>For Plant B
* $x_{2,1}\times5+x_{2,2}\times3\leq60$
* $x_{2,1}\times5+x_{2,2}\times6\leq75$
* $x_{2,1} \geq0$
* $x_{2,2} \geq0$

Constraints Two: Plants can't use more raw material than they have.
>For Plant A:
* $4\times(x_{1,1} + x_{1,2})\leq75$

>For Plant B:
* $4\times(x_{2,1} + x_{2,2})\leq45$


## Exercise 2 - Question 2

In [77]:
# Definition of the Variables
x11 = cvx.Variable(1, integer=True)# Plant A Standard
x12 = cvx.Variable(1, integer=True)# Plant A Deluxe
x21 = cvx.Variable(1, integer=True)# Plant B Standard
x22 = cvx.Variable(1, integer=True)# Plant B Deluxe

# Constraints for Plant A and Plant B
c1 = x11*4+x12*2<=80
c2 = x11*2+x12*5<=60
c3 = x21*5+x22*3<=60
c4 = x21*5+x22*6<=75
c5 = x11>=0
c6 = x12>=0
c7 = x21>=0
c8 = x22>=0
c9 = x11*4+x12*4<=75
c10 = x21*4+x22*4<=45

con_A = [c1, c2, c5, c6, c9]
con_B = [c3, c4, c7, c8, c10]

# Objective function for Plant A and Plant B
OF_A = 10*x11 + 15*x12
OF_B = 10*x21 + 15*x22
objective_A = cvx.Maximize(OF_A)
objective_B = cvx.Maximize(OF_B)

#solving
result_A = cvx.Problem(objective_A, con_A).solve()
result_B = cvx.Problem(objective_B, con_B).solve()

print('optimal profit for Plant A:', result_A)
print('Decision variables')
print("x11 is: ", x11.value)
print("x12 is: ", x12.value)

print('optimal profit for Plant B:', result_B)
print('Decision variables')
print("x21 is: ", x21.value)
print("x22 is: ", x22.value)

optimal profit for Plant A: 220.0
Decision variables
x11 is:  [10.]
x12 is:  [8.]
optimal profit for Plant B: 165.0
Decision variables
x21 is:  [0.]
x22 is:  [11.]


## Exercise 2 - Question 3
Decision variables: $x_{i,j}$.

Let $x_{i,j}$ be plant $i$ makes $x$ amount of $j$ products. 
* $x$ is numerical. 
* $i\  \epsilon\ \{1,2\}$, representing Plant A and Plant B.
* $j\  \epsilon\ \{1, 2\}$, representing Standard and Deluxe.

Goal: Assuming to maximize the profit.

Objective function:
* maximize: $10\times(x_{1,1}+x_{2,1})+15\times(x_{1,2}+x_{2,2})$

Constraints One: the time to produce products can't exceed the plant's capacity.
* $x_{1,1}\times4+x_{1,2}\times2\leq80$
* $x_{1,1}\times2+x_{1,2}\times5\leq60$
* $x_{2,1}\times5+x_{2,2}\times3\leq60$
* $x_{2,1}\times5+x_{2,2}\times6\leq75$
* $x_{1,1} \geq0$
* $x_{1,2} \geq0$
* $x_{2,1} \geq0$
* $x_{2,2} \geq0$

Constraints Two: Plants can't use more raw material than they have.
* $4\times(x_{1,1} + x_{1,2})\leq75$
* $4\times(x_{2,1} + x_{2,2})\leq45$


In [78]:
# Constraints
con = [c1,c2,c3,c4,c5,c6,c7,c8,c9,c10]

#objective function
OF = 10*(x11+x21)+15*(x12+x22)
objective = cvx.Maximize(OF)

#solving
prob = cvx.Problem(objective, con)
result = prob.solve()

print('optimal profit')
print(prob.value)

print('Decision variables')
print("x11 is: ", x11.value)
print("x12 is: ", x12.value)
print("x21 is: ", x21.value)
print("x22 is: ", x22.value)

optimal profit
385.0
Decision variables
x11 is:  [10.]
x12 is:  [8.]
x21 is:  [0.]
x22 is:  [11.]


## Exercise 2 - Question 4
The results are the same as question 2's, because Plant A and B operate individually, and they don't have influence over each other.

## Exercise 2 - Question 5
(For detailed analysis, please refer to the next block.)
To excavate higher potential profit, we should consider rebalancing the allocation of raw material. If Plant A can have 72kg of raw material and Plant B have 48 kg, the optimal profit increase from $385 to $400 with no additional cost. The extra profit results from the extra Deluxe produced by Plant B.

In [79]:
# Constraints
c11 = 4*(x11+x12+x21+x22)<=120
con = [c1,c2,c3,c4,c5,c6,c7,c8,c11]

#objective function
OF = 10*(x11+x21)+15*(x12+x22)
objective = cvx.Maximize(OF)

#solving
prob = cvx.Problem(objective, con)
result = prob.solve()

print('optimal profit')
print(prob.value)

print('Decision variables')
print("x11 is: ", x11.value)
print("x12 is: ", x12.value)
print("x21 is: ", x21.value)
print("x22 is: ", x22.value)
print(f"Plant A needs {4*(x11.value+x12.value)}kg of raw material.")
print(f"Plant A needs {4*(x21.value+x22.value)}kg of raw material.")

optimal profit
400.0
Decision variables
x11 is:  [10.]
x12 is:  [8.]
x21 is:  [0.]
x22 is:  [12.]
Plant A needs [72.]kg of raw material.
Plant A needs [48.]kg of raw material.


## Exercise 3 - Question 1
Variables positively affect profitability:
1. Price of the Inn;
2. College Students in the Area.

Variables negatively affect profitability:
1. State Population per inn;
2. Square Root of the Median Income in the area.

The only variable that doesn't make sense is the state population per inn, as it negatively affects the profitability. The larger the population per inn, the market should lean more toward the seller's market. LQM should be able to price the hotel at a higher price and earn more profit.

One thing worths pointing out is the median income. If LQM were mainly focus on low-end hotels, it's possible that the coefficient is negative as rich people tend to stay less often. 

## Exercise 3 - Question 2

In [80]:
import pandas as pd
import numpy as np


# load the excel file
URL = ""
df = pd.read_excel(URL, header=2, index_col=0)

# add intercept as per the model's requirement
intercept = 1
df['intercept'] = intercept
price = df['Price']
df = df[['intercept', 'State Population Per Inn (normalized)', 'Price (normalized)', 'Square Root of Median Income (normalized)', 'College Students in Area (normalized)']]
df.head()

intercept  ...  College Students in Area (normalized)
Hotel             ...                                       
1              1  ...                              -0.536413
2              1  ...                               0.311669
3              1  ...                               0.311669
4              1  ...                               0.311669
5              1  ...                               0.311669

[5 rows x 5 columns]

In [81]:
# coefficients of the model
profitability_equation = [39.05, -5.41, 5.86, -3.09, 1.75]
profitability = round(df.dot(profitability_equation), 2)
print(profitability)
print(f"Hotel with highest profitability is hotel {profitability[profitability==max(profitability)].index.values}, profit is {max(profitability)}.")
print(f"Hotel with highest profitability is hotel {profitability[profitability==min(profitability)].index.values}, profit is {min(profitability)}.")

Hotel
1     44.24
2     53.38
3     43.02
4     42.61
5     37.34
6     49.10
7     23.78
8     23.45
9     28.67
10    38.88
11    38.01
12    40.29
13    39.42
14    42.36
15    38.59
16    37.39
dtype: float64
Hotel with highest profitability is hotel [2], profit is 53.38.
Hotel with highest profitability is hotel [8], profit is 23.45.


## Exercise 3 - Question 3.a
Adopting such approach, LQM will only purchase Hotel 2, because the price equals to its budget and the second most profitable hotel is not free.

## Exercise 3 - Question 3.b
The total predicted profitability is 53.38 (unit unknown).

## Exercise 3 - Question 3.c
It's not a good approach. From the investment's standpoint, the maximum return will achieve when LQM buys hotels with the highest return on investments (ROI). Buying hotels with the highest value of return and omit the amount of investment, it's not an optimal strategy.

Maximizing the average predicted profitability is an equally bad idea, as it's still not optimizing the ROI.

Bumping up the budget $20 million will alleviate the issue because compared to before, we will be buying hotels with higher ROI than hotel 2. However, the major problem remains because we are still not optimizing ROI.

## Exercise 3 - Question 4.a

In [82]:
# Decision variables.
# Let x be the boolean decision of whether to purchase hotel 1 through 16.
x = cvx.Variable(16, boolean=True)

# Contraint: can't exceed the budget
c = cvx.sum(cvx.multiply(x, np.array(list(price)))) <= 10000000
con = [c]

# Optimization function, maximize profitability
OF = cvx.sum(cvx.multiply(x, np.array(list(profitability))))
objective = cvx.Maximize(OF)

#solving
result = cvx.Problem(objective, con).solve()

print('optimal profitability:', result)
print("x is: ", x.value)
print("Hotel to purchase:")
for count, ele in enumerate(x.value):
  if int(ele) == 1:
    print(df.index[count])

optimal profitability: 269.92
x is:  [0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 1. 1. 0. 1. 1.]
Hotel to purchase:
5
10
11
12
13
15
16


## Exercise 3 - Question 4.b
When budget is capped at $10 million, the best strategy is to purchase hotel 5, 10, 11, 12, 13, 15, and 16. Profitability is 269.92.

## Exercise 3 - Question 4.c
The optimal solution makes intuitive sense. The portfolio is aiming to buy hotels with high ROI, not necessarily the ones with the highest value of profit. With the same amount of budget, the profitability increased to 269.92, compared with 53.38 from the greedy solution.

## Exercise 3 - Question 5.a
let $x_i$ be the purchase decision in city $i$.
* $i\ \epsilon\ \{1, 2, 3, 4, 5\}$, representing Eureka, Fresno, Long Beach, Los Angeles, and South Lake Tahoe.

Additional constraints: can't buy more than two hotels in one city.
* sum($x_2$) $\leq2$
* sum($x_4$) $\leq2$
* sum($x_5$) $\leq2$

I expect the new optimal value to be smaller than or equal to the original optimal value, because it's adding new constraints to the previous situation.

## Exercise 3 - Question 5.b

In [83]:
# Decision variables.
# xi is the decision of whether to purchase any of the hotels in city i.
x1 = cvx.Variable(1, boolean=True)
x2 = cvx.Variable(4, boolean=True)
x3 = cvx.Variable(1, boolean=True)
x4 = cvx.Variable(3, boolean=True)
x5 = cvx.Variable(7, boolean=True)

# Prices of hotels in city i.
p1 = np.array(list(price))[0]
p2 = np.array(list(price))[1:5]
p3 = np.array(list(price))[5]
p4 = np.array(list(price))[6:9]
p5 = np.array(list(price))[9:16]

# Constraint One: can't exceed the budget
c1 = x1*p1 + cvx.sum(cvx.multiply(x2, p2)) + x3*p3 + cvx.sum(cvx.multiply(x4, p4)) + cvx.sum(cvx.multiply(x5, p5)) <= 10000000

# Constraints Two: no more than two hotels in one city
c2 = cvx.sum(x2) <= 2
c3 = cvx.sum(x4) <= 2
c4 = cvx.sum(x5) <= 2
con = [c1,c2,c3,c4]

# Optimization function, maximize profitability
profitability1 = np.array(list(profitability))[0]
profitability2 = np.array(list(profitability))[1:5]
profitability3 = np.array(list(profitability))[5]
profitability4 = np.array(list(profitability))[6:9]
profitability5 = np.array(list(profitability))[9:16]
OF = x1*profitability1 + cvx.sum(cvx.multiply(x2, profitability2)) + x3*profitability3 + cvx.sum(cvx.multiply(x4, profitability4)) + cvx.sum(cvx.multiply(x5, profitability5))
objective = cvx.Maximize(OF)

#solving
result = cvx.Problem(objective, con).solve()

print('optimal profitability:', result)
x = x1.value.tolist()+x2.value.tolist()+x3.value.tolist()+x4.value.tolist()+x5.value.tolist()
print("x is: ", x)
print("Hotel to purchase:")
for count, ele in enumerate(x):
  if int(ele) == 1:
    print(df.index[count])

optimal profitability: 205.7
x is:  [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 1.0, 0.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0]
Hotel to purchase:
1
5
7
8
10
11


## Exercise 3 - Question 5.c
The optimal profitability is 205.7, LQM should buy Hotel 1, 5, 7, 8, 10, and 11.

Compared with the previous profitability of 269.92, the new optimal profitability decreased because the purchasing decision was changed.

## Exercise 3 - Question 6
Given a $10 million budget and based on our already established profit estimation model, the optimal profitability will be 205.7. However, if we remove the 2 hotels per city rule and bet more on South Lake Tahoe, our profitability will have more space to increase. When buying 6 hotels in Tahoe, the profitability will peak at 269.92. Also, I see even more opportunities in optimizing our profit. Our previous profit estimation model is for low-end hotels, as we earn less in richer cities. We can segment the hotels and create a separate model for high-end hotels. It will enable us to make cleaver decisions in acquiring high-end hotels.